networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/rpc/RpcHandler.java

```diff
@@ -72,24 +90,35 @@ public RpcHandler(final AsyncRunner asyncRunner, RpcMethod rpcMethod) {
             asyncRunner.getDelayedFuture(TIMEOUT),
             () ->
                 new StreamTimeoutException(
-                    "Timed out waiting to initialize stream for method " + rpcMethod.getId()));
+                    "Timed out waiting to initialize stream for protocol(s): "
+                        + String.join(",", rpcMethod.getIds())));
 
     return SafeFuture.notInterrupted(closeInterruptor)
         .thenApply(__ -> connection.muxerSession().createStream(this))
         // waiting for a stream or interrupt
         .thenWaitFor(StreamPromise::getStream)
         .orInterrupt(closeInterruptor, timeoutInterruptor)
         .thenCompose(
-            streamPromise ->
-                // waiting for controller, writing initial payload or interrupt
-                SafeFuture.of(streamPromise.getController())
-                    .orInterrupt(closeInterruptor, timeoutInterruptor)
-                    .thenPeek(ctr -> ctr.setRequestHandler(handler))
-                    .thenApply(Controller::getRpcStream)
-                    .thenWaitFor(rpcStream -> rpcStream.writeBytes(initialPayload))
-                    .orInterrupt(closeInterruptor, timeoutInterruptor)
-                    // closing the stream in case of any errors or interruption
-                    .whenException(err -> closeStreamAbruptly(streamPromise.getStream().join())))
+            streamPromise -> {
+              final SafeFuture<String> protocolIdFuture =
+                  SafeFuture.of(streamPromise.getStream().thenCompose(Stream::getProtocol));
+              // waiting for controller, writing initial payload or interrupt
+              return SafeFuture.of(streamPromise.getController())
+                  .thenCombine(
```

Comment:
Can skip the cast below and use:
```suggestion
                  .<String, RpcStreamController<TOutgoingHandler>>thenCombine(
```

Reply:
Thanks - I wasn't sure the right way to fix these casting issues :)